File: owin_fsharp/Tests/ContractTests.fs
namespace Tests
open System
open NUnit.Framework
open FsUnit
open Customers
open System.Text
open FluentAssertions
open FluentAssertions.Xml
open System.Xml.Linq
open Microsoft.Owin.Testing
open System.Net.Http
open System.IO

    [<TestFixture>] 
    type ``ContractTests``() =
        let customer = {Customer.Empty with AccountNumber = 1; FirstName = "Oskar"; LastName = "Gewalli" }
        let ns =  "xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xmlns:xsd=\"http://www.w3.org/2001/XMLSchema\" xmlns=\"http://schemas.datacontract.org/2004/07/Customers\""
        let serializedCustomers = String.Format( @"<ArrayOfCustomer {0}>
  <Customer>
    <AccountNumber>1</AccountNumber>
    <AddressCity xsi:nil=""true"" />
    <AddressCountry xsi:nil=""true"" />
    <AddressStreet xsi:nil=""true"" />
    <FirstName>Oskar</FirstName>
    <Gender>Male</Gender>
    <LastName>Gewalli</LastName>
    <PictureUri xsi:nil=""true"" />
  </Customer>
</ArrayOfCustomer>", ns)

        let serialized customer =
            let buffer = Serializer.serialize( CustomerOutput.Single(customer) )
            let s = new MemoryStream()
            let w = new StreamWriter(s)
            w.Write(buffer)
            w.Flush()
            s.Seek(0L, SeekOrigin.Begin) |> ignore
            s

        let should_be_xml_equivalent expected actual =
            ignore(XDocument.Parse(actual).Should()
                    .BeEquivalentTo(XDocument.Parse(expected)))

        let context c=
            let svc = new CustomerServiceFake(c)
            TestServer.Create(fun b-> 
                let a = new HttpAdapter(svc)
                a.configuration(b)
                )
 
        let GETXml (server: TestServer) (url:string)=
            server.HttpClient.GetAsync(url).Result.Content.ReadAsStringAsync().Result;
        let POSTXml (server: TestServer) (url:string) c=
            server.HttpClient.PostAsync(url, c).Result.Content.ReadAsStringAsync().Result;



        [<Test>] member test.
         ``get all customers`` ()=
            let result = GETXml (context([|customer|])) "/CustomerService.svc/GetAllCustomers" in
                should_be_xml_equivalent serializedCustomers result
                    
        [<Test>] member test.
         ``save customer`` ()=
            using (new StreamContent(serialized customer))
            (fun s->
                let result = POSTXml (context([|customer|])) "/CustomerService.svc/SaveCustomer" s 
                should_be_xml_equivalent @"<boolean>true</boolean>" result
            )

File: owin/Tests/ContractTests.cs
using System;
using Xunit;
using Customers;
using FluentAssertions;
using System.Xml.Linq;
using System.IO;
using System.Text;
using Microsoft.Owin.Testing;
using System.Net.Http;
using System.Linq;

namespace Tests
{
    public class ContractTests:IDisposable
    {
        private TestServer adapter;
        private CustomerServiceFake svc;

        public ContractTests()
        {
            svc = new CustomerServiceFake();
            adapter = TestServer.Create(b=> new Startup(new Serializer(), svc).Configuration(b));
        }
        public void Dispose()
        {
            adapter.Dispose();
        }

        [Fact]
        public void GetAllCustomers()
        {
            svc.AllCustomers.Add(new Customer
                { 
                    AccountNumber = 1, 
                    FirstName = "Oskar", 
                    LastName = "Gewalli" 
                });
            var result = adapter.HttpClient.GetAsync("/CustomerService.svc/GetAllCustomers").Result;
            XDocument.Parse(result.Content.ReadAsStringAsync().Result).Should().BeEquivalentTo(XDocument.Parse(@"<?xml version=""1.0"" encoding=""utf-8""?>
<ArrayOfCustomer xmlns:xsi=""http://www.w3.org/2001/XMLSchema-instance"" xmlns:xsd=""http://www.w3.org/2001/XMLSchema"" xmlns=""http://schemas.datacontract.org/2004/07/Customers"">
  <Customer>
    <AccountNumber>1</AccountNumber>
    <AddressCity xsi:nil=""true"" />
    <AddressCountry xsi:nil=""true"" />
    <AddressStreet xsi:nil=""true"" />
    <FirstName>Oskar</FirstName>
    <Gender>Male</Gender>
    <LastName>Gewalli</LastName>
    <PictureUri xsi:nil=""true"" />
  </Customer>
</ArrayOfCustomer>"));
        }

        [Fact]
        public void SaveCustomer()
        {
            svc.AllCustomers.Add(new Customer
                { 
                    AccountNumber = 1, 
                    FirstName = "Oskar", 
                    LastName = "Gewalli" 
                });
            using (var c = AsStream(new Customer
                { 
                    AccountNumber = 1, 
                    FirstName = "Oskar", 
                    LastName = "GewalliZ" 
                }))
            {
                var result = adapter.HttpClient.PostAsync("/CustomerService.svc/SaveCustomer", new StreamContent(c)).Result;
                XDocument.Parse(result.Content.ReadAsStringAsync().Result).Should().BeEquivalentTo(XDocument.Parse(@"<?xml version=""1.0"" encoding=""utf-8""?>
<boolean>true</boolean>"));
                Assert.Equal("GewalliZ", svc.AllCustomers.Single().LastName);
            }
        }

        Stream AsStream(Customer customer)
        {
            var serializer = new Serializer();
            var stream = new MemoryStream();
            var writer=new StreamWriter(stream);
            var buffer = serializer.Serialize(customer);
            writer.Write(buffer);
            writer.Flush();
            stream.Seek(0, SeekOrigin.Begin);
            return stream;
        }
    }
}

